library/core/src/hash/mod.rs

```diff
@@ -153,9 +153,19 @@ mod sip;
 /// Thankfully, you won't need to worry about upholding this property when
 /// deriving both [`Eq`] and `Hash` with `#[derive(PartialEq, Eq, Hash)]`.
 ///
+/// ## Prefix collisions
+///
+/// Implementations of `hash` should ensure that the data they
+/// pass to the `Hasher` are prefix-free. That is, different concatenations
```

Comment:
I this this explanation of what "prefix-free" means is not accurate or incomplete. It should say that unequal values should cause two different byte sequences to be written, and neither of the two sequences should be a prefix of the other.

Note that it's not sufficient to say that concatenations of multiple values of the type should result in different outputs. It has to be true when concatenated with outputs for other types as well. Which is true if all these types satisfy the prefix-free property.

Reply:
Thanks @tczajka! I'm not sure I understand the idea of one sequence being a prefix of another. Does it simply mean "starts with", or is it another kind of relation? Is there a way we can rephrase this?

Reply:
Another way to ask the question: in the example of `("ab", "c")` and `("a", "bc")` where and how would the "prefix" occur, and how does the extra byte prevent it?

Reply:
A "prefix" is a beginning of a string, so it's same as "starts with". https://en.wikipedia.org/wiki/Prefix

If strings were hashed without the extra 0xff at the end, hashing `("ab", "c")` and `("a", "bc")` would write the same byte sequence `"abc"` to `Hasher`. The problem is that `"a"` is a prefix of `"ab"`. Whereas `"a\xff"` is not a prefix of `"ab\xff"`, so if `Hash` outputs these sequences instead that solves the problem. `"ab\xffc\xff" != "a\xffbc\xff"`.

Reply:
Note: `\xff` is not actually allowed in string literals, since it would be invalid UTF-8 -- which is also what makes it a useful separator here.